### ISSUE TITLE
[aws-dynamodb] tableStreamArn is undefined

### ISSUE THREAD

**Issue by phekno** [aws/aws-cdk#9466]
We're creating a DynamoDB table with a stream.  We're trying to create a `lambda.EventSourceMapping`, passing in `table.tableStreamArn` as the `eventSourceArn` prop like so:
```
    new lambda.EventSourceMapping(this, props.serviceName + 'eventSourceMapping', {
      eventSourceArn: ddbTable.tableStreamArn,
      target: ddbStreamProcessor,
      startingPosition: lambda.StartingPosition.TRIM_HORIZON
    });
```
When we try to build (transpiling from TypeScript to JavaScript), we get the following error:
```
npm run build

> cdk@0.1.0 build /home/ec2-user/scm/app/cdk
> tsc

lib/cdk-stack.ts:63:7 - error TS2322: Type 'string | undefined' is not assignable to type 'string'.
  Type 'undefined' is not assignable to type 'string'.

63       eventSourceArn: ddbTable.tableStreamArn,
         ~~~~~~~~~~~~~~

  node_modules/@aws-cdk/aws-lambda/lib/event-source-mapping.d.ts:9:14
    9     readonly eventSourceArn: string;
                   ~~~~~~~~~~~~~~
    The expected type comes from property 'eventSourceArn' which is declared here on type 'EventSourceMappingProps'


Found 1 error.
```

### Environment
  - **CDK CLI Version:** `1.56.0 (build c1c174d)`
  - **Module Version:** `1.56.0`
  - **Node.js Version:** `v10.16.0`
  - **OS:** all
  - **Language (Version):** TypeScript 3.7.2

**Comment by phekno** [comment]
I was incorrect in saying that `cdk synth` worked fine.  It was just rendering the template based on a previously built JS file.

**Comment by RomainMuller** [comment]
The `tableStreamArn` property is optional (meaning it can be `undefined`) because no stream will exist if the table was created without a `stream` property. This is perfectly normal and expected.

There are two things you can do to fix your code:
1. If you're certain the table was created with a `stream` property, use a `!` to inform the typescript compiler that you **know** the value will not be `undefined`:
   ```ts
   eventSourceArn: ddbTable.tableStreamArn!,
   ```
2. If you are obtaining an `ITable` from outside of your control and cannot guarantee it was created with a `stream` property, add guard for the case where no stream was created:
   ```ts
   if (ddbTable.tableStreamArn == null) {
      throw new Exception("No table stream is available on this DynamoDB table. It is required."); // Or whatever
   }
   // From here on, the typescript compiler **knows** `ddbTable.tableStreamArn` cannot be `undefined`
   // since an exception would have been thrown if that was the case. It won't complain about the incompatible
   // types in assignment anymore.
   ```

**Comment by 53casey** [comment]
This might be the best place to post the question about an issue I am facing. I am doing a cross-stack cross-stack deployment with one stack containing CDK files for S3, IAM, Dynamodb, Codecommit, Cloudfront, etc. while individual stacks are created for each Lambda. Now I somehow need to use the Dynamodb stream created in one stack as an event trigger for the lambda.

Dynamodb table is built like this:
```
const someTableName = new dynamodb.Table(this, 'someTableName', {
      partitionKey: {
        name: 'Field_ID_1',
        type: dynamodb.AttributeType.STRING
      },
      stream: dynamodb.StreamViewType.NEW_IMAGE,
      tableName: 'someTableName',
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      removalPolicy: cdk.RemovalPolicy.DESTROY
    });

    someTableName.addGlobalSecondaryIndex({
      indexName: 'someSecondaryIndex',
      partitionKey: { name: 'Field_ID_2', type: dynamodb.AttributeType.STRING }
    });

    new cdk.CfnOutput(this, 'someTableName_output', {
      value: VehicleStatic.tableArn,
      exportName: 'dynamodb-table-arn-someTableName'
    });
```

I tried 2 approaches in Lambda CDK:

1.

```
const table = dynamodb.Table.fromTableAttributes(this, 'someTableNameIAmImporting', {
            tableArn: cdk.Fn.importValue('dynamodb-table-arn-someTableName')
        });
const tablename = table.tableName
console.log("this is tablename:" + tablename) // This outputs only: this is tablename:${Token[TOKEN.93]}

func.addEventSource(new sources.DynamoEventSource(table, {
        batchSize: 1,
        parallelizationFactor: 1,
        maxRecordAge: cdk.Duration.seconds(0),
        retryAttempts: 0,
        bisectBatchOnError: false,
        startingPosition: lambda.StartingPosition.LATEST
}));
```

2.

```
new lambda.EventSourceMapping(this, 'DynamoDBEventSource', {
        target: func,
        eventSourceArn: cdk.Fn.importValue('dynamodb-table-arn-someTableName'),
        // Also tried, but kept getting an error with:
        // eventSourceArn: cdk.Fn.importValue('dynamodb-table-arn-someTableName').tableStreamArn,
        batchSize: 1,
        parallelizationFactor: 1,
        bisectBatchOnError: false,
        startingPosition: lambda.StartingPosition.LATEST
    });
```

I understand that the Dynamodb Stream takes time to be created so I can't export its Arn at the time I am running my Dynamodb CDK. But I have not wrapped my head around the whole thing yet. I'd love to get a solution to fix this and move on, but I would truly appreciate an explanation because I am definitely missing something when trying cross-stack referencing in order to facilitate Dynamodb Stream to work as a trigger for Lambda.

Thanks in advance!

**Comment by skinny85** [comment]
@53casey can you try passing the Table object into the Stacks with the Lambdas, and referencing it directly?

**Comment by 53casey** [comment]
Hello Adam! First of all thanks for a great blog post (CDK tips, part 3) on cross-stack referencing! It helped me a lot 2 weeks ago when I had to do cross-stack referencing of resources within the same CDK application. This time, however, I am trying to reference resources from different CDK applications. Maybe I didn't make this clear, it's been a very long day so I apologize for any confusion in my original question.

_To return to your response:
"can you try passing the Table object into the Stacks with the Lambdas, and referencing it directly?"_

Maybe I am not understanding you.. I thought that is what I was doing, "referencing resources directly"..
Basically, in my CDK app 1 I create a Dynamodb table and its Stream. But I can't export an Arn of that stream. I can only export an Arn of the table.
Then, in the CDK app 2 I create a Lambda, but when I try to somehow reference the Dynamodb Stream - I can't seem to grab it.

Thank you for your patience.

**Comment by skinny85** [comment]
Hello!

When I say "directly", I mean the same way [my blog post that you mentioned](https://www.endoflineblog.com/cdk-tips-03-how-to-unblock-cross-stack-references) is passing an S3 Bucket from the `ProducingStack` to the `ConsumerStack`.
